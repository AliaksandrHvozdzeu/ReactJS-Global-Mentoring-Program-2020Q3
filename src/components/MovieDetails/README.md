#Movie details in poster component
