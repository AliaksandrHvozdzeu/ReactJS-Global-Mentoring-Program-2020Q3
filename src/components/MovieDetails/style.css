.movie-preview-description-block {
  color: var(--white);
  font-weight: 300;
  margin-left: 1rem;
}

.movie-preview-description *:not(:last-child) {
  margin-bottom: 1rem;
}

.movie-preview-title {
  font-size: 3rem;
  font-weight: 300;
  display: inline-block;
  margin-right: 1rem;
}

.movie-preview-release, .movie-preview-runtime {
  display: inline-block;
  color: var(--red);
  font-size: 1.5rem;
  font-weight: 300;
  position: relative;
  margin-right: 2rem;
  margin-bottom: 1.5rem;
}

.movie-preview-runtime:after {
  content: 'min';
  position: absolute;
  right: -3rem;
}

.movie-preview-description {
  width: 100%;
}

.movie-preview-rating {
  font-size: 1.5rem;
  color: #30ea30;
  border-radius: 50%;
  padding: 1rem;
  border: 2px solid var(--grey);
  display: inline-block;
  transform: translateY(-5px);
}

.movie-preview-rating-yellow {
  font-size: 1.5rem;
  color: #eee50f;
  border-radius: 50%;
  padding: 1rem;
  border: 2px solid var(--grey);
  display: inline-block;
  transform: translateY(-5px);
}

.movie-preview-rating-red {
  font-size: 1.5rem;
  color: var(--red);
  border-radius: 50%;
  padding: 1rem;
  border: 2px solid var(--grey);
  display: inline-block;
  transform: translateY(-5px);
}

.movie-preview-genres {
  color: var(--white);
  margin-bottom: 1.5rem;
}
