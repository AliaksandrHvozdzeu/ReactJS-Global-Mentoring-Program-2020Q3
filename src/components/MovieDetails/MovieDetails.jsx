import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

const MovieDetails = ({ details }) => (
  <div className="movie-preview-description-block">
    <h1 className="movie-preview-title">{details.title}</h1>
    <p className="movie-preview-rating">{details.rating}</p>
    <p className="movie-preview-genres">{details.genres.join(', ')}</p>
    <p className="movie-preview-release">{new Date(details.release).getFullYear()}</p>
    <p className="movie-preview-runtime">{details.runtime}</p>
    <p className="movie-preview-description">{details.description}</p>
  </div>
);

MovieDetails.propTypes = {
  details: PropTypes.shape({
    title: PropTypes.string,
    rating: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    release: PropTypes.string,
    runtime: PropTypes.number,
    description: PropTypes.string,
  }).isRequired,
};

export default MovieDetails;
