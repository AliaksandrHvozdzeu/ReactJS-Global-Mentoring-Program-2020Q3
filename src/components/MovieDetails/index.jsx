import React from 'react';
import PropTypes from 'prop-types';
import './style.css';

function ratingGradient(rating) {
  return rating >= 3 && rating <= 3.5 ? <p className="movie-preview-rating-yellow">{rating}</p> : rating < 3 ?
    <p className="movie-preview-rating-red">{rating}</p> : <p className="movie-preview-rating">{rating}</p>;
}

const MovieDetails = ({ details }) => (
  <div className="movie-preview-description-block">
    <h1 className="movie-preview-title">{details.title}</h1>
    {ratingGradient(details.rating)}
    <p className="movie-preview-genres">{details.genres.join(', ')}</p>
    <p className="movie-preview-release">{new Date(details.release).getFullYear()}</p>
    <p className="movie-preview-runtime">{details.runtime}</p>
    <p className="movie-preview-description">{details.description}</p>
  </div>
);

MovieDetails.propTypes = {
  details: PropTypes.shape({
    title: PropTypes.string.isRequired,
    rating: PropTypes.string.isRequired,
    genres: PropTypes.arrayOf(PropTypes.string),
    release: PropTypes.string.isRequired,
    runtime: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
  }),
};

export default MovieDetails;
