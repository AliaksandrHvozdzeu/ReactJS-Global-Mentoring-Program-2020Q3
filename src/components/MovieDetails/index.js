import MovieDetails from './js/MovieDetails';

export default MovieDetails;
