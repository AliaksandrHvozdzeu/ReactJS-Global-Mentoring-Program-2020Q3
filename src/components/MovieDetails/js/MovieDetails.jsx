import React, { useState } from 'react';
import PropTypes from 'prop-types';
import defaultPoster from '../../../../public/images/not-found.png';
import MoviePoster from '../../MoviePoster';
import PosterPreview from '../../PosterPreview/js/PosterPreview';
import '../css/MovieDetails.css';

export default function MovieDetails({ details }) {

  const [isOpen, setIsOpen] = useState(false);

  const defaultVoteAverage = (voteAverage) => {
    return voteAverage === undefined ? '0.0' : voteAverage;
  };

  const openModal = () => {
    setIsOpen(true);
  };

  const closeModal = () => {
    setIsOpen(false);
  };

  return details == null ? (
    <div className="movie-details-not-found">
      Movie was not found, choose another.
    </div>
  ) : (
    <>
      <MoviePoster src={details.poster_path} alt="Movie poster" fallback={defaultPoster} onClick={openModal}/>
      <div className="movie-preview-description-block">
        <h1 className="movie-preview-title">{details.title}</h1>
        <p className="movie-preview-rating">{defaultVoteAverage(details.vote_average)}</p>
        <p className="movie-preview-genres">{details.genres.join(', ')}</p>
        <p className="movie-preview-release">{new Date(details.release_date).getFullYear()}</p>
        <p className="movie-preview-runtime">{details.runtime}</p>
        <p className="movie-preview-description">{details.overview}</p>
      </div>
      <PosterPreview isOpen={isOpen}
                     onClose={closeModal}
                     src={details.poster_path}
                     alt="movie poster preview"
                     fallback={defaultPoster}/>
    </>
  );

}

MovieDetails.propTypes = {
  details: PropTypes.shape({
    title: PropTypes.string,
    vote_average: PropTypes.string,
    genres: PropTypes.arrayOf(PropTypes.string),
    poster_path: PropTypes.string,
    release_date: PropTypes.string,
    runtime: PropTypes.number,
    overview: PropTypes.string,
  }).isRequired,
};

