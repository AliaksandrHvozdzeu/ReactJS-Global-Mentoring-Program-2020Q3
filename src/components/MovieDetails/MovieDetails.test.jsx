import React from "react";
import rerender from "react-test-renderer";
import MovieDetails from "./js/MovieDetails";
import { shallow } from "enzyme";

describe("when MovieDetails", () => {

  test("then poster_path is empty", () => {
    const details = {
      "vote_average": "2",
      "poster_path": ""
    };
    const component = rerender.create(<MovieDetails details={details} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("then poster_path is not correct", () => {
    const details = null;
    const component = rerender.create(<MovieDetails details={details} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("then snapshot created when details is null", () => {
    const details = {
      "title": "Test",
      "release_date": "2020-10-21",
      "poster_path": "https://upload.wikimedia.org/wikipedia/commons/4/47/PNG_transparency_demonstration_1.png",
      "genres": ["Comedy", "Crime", "Documentary", "Horror"],
      "overview": "over",
      "runtime": 120
    };
    const component = rerender.create(<MovieDetails details={details} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders", () => {
    const details = {
      "title": "Test",
      "release_date": "2020-10-21",
      "poster_path": "https://upload.wikimedia.org/wikipedia/commons/4/47/PNG_transparency_demonstration_1.png",
      "genres": ["Comedy", "Crime", "Documentary", "Horror"],
      "overview": "over",
      "runtime": 120
    };
    const wrapper = shallow(
      <MovieDetails details={details} />
    );

    expect(wrapper.exists()).toBe(true);
  });

});
