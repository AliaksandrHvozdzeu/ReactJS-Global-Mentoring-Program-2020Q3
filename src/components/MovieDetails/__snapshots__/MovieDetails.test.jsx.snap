// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when MovieDetails then poster_path is empty 1`] = `
Array [
  <img
    alt="Movie poster"
    className="poster"
    onClick={[Function]}
    onError={[Function]}
    role="presentation"
    src=""
  />,
  <div
    className="movie-preview-description-block"
  >
    <h1
      className="movie-preview-title"
    />
    <p
      className="movie-preview-rating"
    >
      2
    </p>
    <p
      className="movie-preview-genres"
    />
    <p
      className="movie-preview-release"
    >
      NaN
    </p>
    <p
      className="movie-preview-runtime"
    />
    <p
      className="movie-preview-description"
    />
  </div>,
]
`;

exports[`when MovieDetails then poster_path is not correct 1`] = `
<div
  className="movie-details-not-found"
>
  Movie was not found, choose another.
</div>
`;

exports[`when MovieDetails then snapshot created when details is null 1`] = `
Array [
  <img
    alt="Movie poster"
    className="poster"
    onClick={[Function]}
    onError={[Function]}
    role="presentation"
    src="https://upload.wikimedia.org/wikipedia/commons/4/47/PNG_transparency_demonstration_1.png"
  />,
  <div
    className="movie-preview-description-block"
  >
    <h1
      className="movie-preview-title"
    >
      Test
    </h1>
    <p
      className="movie-preview-rating"
    >
      0.0
    </p>
    <p
      className="movie-preview-genres"
    />
    <p
      className="movie-preview-release"
    >
      2020
    </p>
    <p
      className="movie-preview-runtime"
    >
      120
    </p>
    <p
      className="movie-preview-description"
    >
      over
    </p>
  </div>,
]
`;
